sycl/source/detail/queue_impl.cpp

```diff
@@ -189,60 +239,57 @@ event queue_impl::memcpy(const std::shared_ptr<detail::queue_impl> &Self,
-  }
+  };
+  if (MGraph.lock())
+    return submitWithScheduler();
+
   if ((!Src || !Dest) && Count != 0) {
     report(CodeLoc);
     throw runtime_error("NULL pointer argument in memory copy operation.",
                         PI_ERROR_INVALID_VALUE);
   }
-  if (MHasDiscardEventsSupport) {
-    MemoryManager::copy_usm(Src, Self, Count, Dest,
-                            getOrWaitEvents(DepEvents, MContext), nullptr);
-    return createDiscardedEvent();
-  }
 
-  event ResEvent = prepareSYCLEventAssociatedWithQueue(Self);
   {
-    // We need to submit command and update the last event under same lock if we
-    // have in-order queue.
-    auto ScopeLock = isInOrder() ? std::unique_lock<std::mutex>(MLastEventMtx)
-                                 : std::unique_lock<std::mutex>();
-    // If the last submitted command in the in-order queue is host_task then
-    // wait for it before submitting usm command.
-    if (isInOrder() && MLastCGType == CG::CGTYPE::CodeplayHostTask)
-      MLastEvent.wait();
+    std::unique_lock<std::mutex> Lock(MMutex, std::defer_lock);
 
     std::vector<event> MutableDepEvents;
     const std::vector<event> &ExpandedDepEvents =
-        getExtendDependencyList(DepEvents, MutableDepEvents);
-
-    auto EventImpl = detail::getSyclObjImpl(ResEvent);
-    MemoryManager::copy_usm(Src, Self, Count, Dest,
-                            getOrWaitEvents(ExpandedDepEvents, MContext),
-                            &EventImpl->getHandleRef(), EventImpl);
-
-    if (MContext->is_host())
-      return MDiscardEvents ? createDiscardedEvent() : event();
-
-    if (isInOrder()) {
-      MLastEvent = ResEvent;
-      // We don't create a command group for usm commands, so set it to None.
-      // This variable is used to perform explicit dependency management when
-      // required.
-      MLastCGType = CG::CGTYPE::None;
+        getExtendDependencyList(DepEvents, MutableDepEvents, Lock);
+
+    if (checkEventsForSchedulerBypass(ExpandedDepEvents, MContext)) {
+      if (MHasDiscardEventsSupport) {
+        MemoryManager::copy_usm(Src, Self, Count, Dest,
+                                getPIEvents(ExpandedDepEvents), nullptr);
+        return createDiscardedEvent();
+      }
+
+      event ResEvent = prepareSYCLEventAssociatedWithQueue(Self);
+      auto EventImpl = detail::getSyclObjImpl(ResEvent);
+      MemoryManager::copy_usm(Src, Self, Count, Dest,
+                              getPIEvents(ExpandedDepEvents),
+                              &EventImpl->getHandleRef(), EventImpl);
+      if (MContext->is_host())
+        return MDiscardEvents ? createDiscardedEvent() : event();
+      if (isInOrder()) {
+        auto &EventToStoreIn =
+            MGraph.lock() ? MGraphLastEventPtr : MLastEventPtr;
+        EventToStoreIn = EventImpl;
+      }
+      // Track only if we won't be able to handle it with piQueueFinish.
+      if (MEmulateOOO)
+        addSharedEvent(ResEvent);
+      return discard_or_return(ResEvent);
```

Comment:
Can we outline it into a helper? I'd prefer the refactoring to land before this PR, unless the fix is really urgent.

Reply:
Done

Reply:
Opened https://github.com/intel/llvm/pull/12474